Core/include/Acts/TrackFitting/Chi2Fitter.hpp

```diff
@@ -0,0 +1,767 @@
+  /// @param mctx The magnetic context for this fit
+  /// @param cctx The calibration context for this fit
+  /// @param extensions_ The chi2 extensions
+  /// @param logger_ The logger wrapper
+  /// @param pOptions The plain propagator options
+  /// @param mScattering Whether to include multiple scattering
+  /// @param eLoss Whether to include energy loss
+  /// @param nIter Number of update steps to the parameters
+  /// @param calcFinalChi2_ Whether to run additional propagation to calculate
+  /// final chi2
+  /// @param freeToBoundCorrection_ Correction for non-linearity effect during
+  /// transform from free to bound
+  Chi2FitterOptions(const GeometryContext& gctx,
+                    const MagneticFieldContext& mctx,
+                    std::reference_wrapper<const CalibrationContext> cctx,
+                    Chi2FitterExtensions<traj_t> extensions_,
+                    LoggerWrapper logger_,
+                    const PropagatorPlainOptions& pOptions,
+                    bool mScattering = false, bool eLoss = false, int nIter = 1,
+                    bool calcFinalChi2_ = true,
+                    const FreeToBoundCorrection& freeToBoundCorrection_ =
+                        FreeToBoundCorrection(false))
+      : geoContext(gctx),
+        magFieldContext(mctx),
+        calibrationContext(cctx),
+        extensions(std::move(extensions_)),
+        propagatorPlainOptions(pOptions),
+        multipleScattering(mScattering),
+        energyLoss(eLoss),
+        nUpdates(nIter),
+        calcFinalChi2(calcFinalChi2_),
+        freeToBoundCorrection(freeToBoundCorrection_),
+        logger(logger_) {}
+  /// Contexts are required and the options must not be default-constructible.
+  Chi2FitterOptions() = delete;
+
+  /// Context object for the geometry
+  std::reference_wrapper<const GeometryContext> geoContext;
+  /// Context object for the magnetic field
+  std::reference_wrapper<const MagneticFieldContext> magFieldContext;
+  /// context object for the calibration
+  std::reference_wrapper<const CalibrationContext> calibrationContext;
+
+  Chi2FitterExtensions<traj_t> extensions;
+
+  /// The trivial propagator options
+  PropagatorPlainOptions propagatorPlainOptions;
+
+  /// Whether to consider multiple scattering
+  bool multipleScattering = false;  // TODO: add later
+
+  /// Whether to consider energy loss
+  bool energyLoss = false;  // TODO: add later
+
+  /// Number of iterations to improve chi2
+  int nUpdates = 1;
+
+  /// Whether to do an additional propagation step, just to get the latest chi2
+  /// value
+  bool calcFinalChi2 = true;
+
+  /// Whether to include non-linear correction during global to local
+  /// transformation
+  FreeToBoundCorrection freeToBoundCorrection;
+
+  /// Logger
+  LoggerWrapper logger;
+};
+
+template <typename traj_t>
+struct Chi2FitterResult {
+  // Fitted states that the actor has handled.
+  std::shared_ptr<traj_t> fittedStates;
```

Comment:
Why are those shared ? Is there some reference counting needed (or is it to make the result copyable) ?

Reply:
That's the default now for all fitters now, I think to make it possible to use only a single trajectory object per event or so...